Pydantic serialization failed with nested fields

<!--Please include [tune], [rllib], [autoscaler] etc. in the issue title if relevant-->

### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):*

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):


```
from typing import Any, Dict, List, Optional

import pytest
from pydantic import BaseModel
from ray.serve.utils import register_custom_serializers

import ray

@pytest.fixture(scope="session")
def start_ray():
    ray.init(ignore_reinit_error=True)
    register_custom_serializers()

def test_serialize_nested_field(start_ray):
    class B(BaseModel):
        v: List[int]

    # this shouldn't error
    B(v=[1])

    @ray.remote
    def func():
        # this shouldn't error
        return B(v=[1])

    ray.get(func.remote())
```

somehow the list type hint get swallowed:
```
E                       ray.exceptions.RayTaskError(ValidationError): ray::func() (pid=70624, ip=192.168.31.141)
E                         File "python/ray/_raylet.pyx", line 502, in ray._raylet.execute_task
E                         File "/Users/simonmo/Desktop/ray/ray/python/ray/serve/tests/test_pydantic_serialization.py", line 180, in func
E                           return B(v=[1])
E                         File "pydantic/main.py", line 391, in pydantic.main.BaseModel.__init__
E                       pydantic.error_wrappers.ValidationError: 1 validation error for B
E                       v
E                         value is not a valid integer (type=type_error.integer)

../../worker.py:1434: RayTaskError(ValidationError)
```

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

--- a/python/ray/serve/tests/test_pydantic_serialization.py
+++ b/python/ray/serve/tests/test_pydantic_serialization.py
@@ -167,7 +167,6 @@ def test_serialize_serve_dataclass(start_ray):
     ray.get(consume.remote(BackendConfig()))
 
 
-@pytest.mark.skip("Fails. https://github.com/ray-project/ray/issues/14960.")
 def test_serialize_nested_field(start_ray):
     class B(BaseModel):
         v: List[int]

--- a/python/ray/serve/utils.py
+++ b/python/ray/serve/utils.py
@@ -384,7 +384,7 @@ def register_custom_serializers():
             pydantic.fields.ModelField,
             serializer=lambda o: {
                 "name": o.name,
-                "type_": o.type_,
+                "type_": o.outer_type_,
                 "class_validators": o.class_validators,
                 "model_config": o.model_config,
                 "default": o.default,

--- a/python/ray/serve/utils.py
+++ b/python/ray/serve/utils.py
@@ -384,6 +384,9 @@ def register_custom_serializers():
             pydantic.fields.ModelField,
             serializer=lambda o: {
                 "name": o.name,
+                # outer_type_ is the original type for ModelFields,
+                # while type_ can be updated later with the nested type
+                # like int for List[int].
                 "type_": o.outer_type_,
                 "class_validators": o.class_validators,
                 "model_config": o.model_config,

